Patch `Element.outerHTML`

TimvdLippe opened the pull request:
I took the instructions of @bicknellr in https://github.com/webcomponents/custom-elements/issues/71#issuecomment-287914808 to implement this patch. I also took the original testsuite for `innerHTML` and adapted it with a `container` element for the `outerHTML` property.

Lastly, I implemented the replace logic if the propertydescriptor does not exist. It takes a new element and a container and moves nodes before itself, before removing itself. PTAL as I am not 100% it is correct.

Fixes #71

TimvdLippe committed:
Patch `Element.outerHTML`

TimvdLippe commented:
There is 1 test failure on os x 10.11 safari 9, but I do not have that browser available to me. @bicknellr are you able to run that browser?

bicknellr commented:
Looks like there is a bug Safari 9 where setting `textContent` to `""` actually causes the node's child nodes to be replaced with a single text node with text `""`, rather than just removing all the child nodes:
```javascript
div = document.createElement('div');
div.childNodes.length === 0; // true everywhere
div.textContent = "";
div.childNodes.length; // 1 in Safari 9, 0 everywhere else
```
I'd say change the values that you're setting to `textContent` to something other than the empty string; then, you can assert `div.childNodes.length === 1` after setting `textContent` in all cases.

TimvdLippe committed:
Merge branch 'master' into outerhtml

TimvdLippe committed:
Do not trigger disconncect twice in custom outerHTML patch

Because `container.innerHTML` already calls disconnect
for all nodes, we should not call disconnect for ourselves
again. This should fix the test failure on Safari 9

TimvdLippe commented:
@bicknellr I fixed the Safari 9 issue :tada: (see last commit). Basically, in Safari 9 neither `Element` and `HTMLElement` have `outerHTML`, so the custom patch was run. In the custom patch, I was calling `disconnectTree`. However, `container.innerHTML` already takes care of that. Therefore, disconnects were running twice. Removing the `disconnectTree` in `outerHTML` resolved that situation.

bicknellr reviewed (commented):
I think we need to check that this doesn't have a similar issue as #141. Particularly, if `patchCallback` calls one of the patching functions on the context element's parent node, then I think it might upgrade other children of that parent node earlier than it should.

bicknellr commented on src/Patch/Element.js:
```diff
@@ -116,6 +125,44 @@ export default function(internals) {
+          return Native.Node_cloneNode.call(this, true).outerHTML;
+        },
+        set: /** @this {Element} */ function(assignedValue) {
+          const container = Native.Document_createElement.call(document, 'div');
```
Or maybe `createElementNS(this.parentNode.namespaceURI, ...`?

TimvdLippe replied:
In this case it is not necessary, as `div` is always in the HTML namespace :smile:

bicknellr replied:
True :) , more of what I'm getting at though is that this seems like this would need to do something like [this segment from the `innerHTML` wrapper](https://github.com/webcomponents/custom-elements/blob/a198647cc0877a30c4ab930cf8b0d3ba8e485ad7/src/Patch/Element.js#L103-L105) so that `outerHTML` creates elements in the namespace of the parent.

TimvdLippe commented on src/Patch/Element.js:
```diff
@@ -63,15 +66,21 @@ export default function(internals) {
     });
   }
 
+  function patch_innerHTML(destination, baseDescriptor) {
```
This would generate code duplication in three different places. I would prefer not to.

TimvdLippe committed:
Change ordering of parameters of patch_HTMLsetter

TimvdLippe committed:
Merge branch 'master' into outerhtml

TimvdLippe committed:
Fix imports in outerHTML test

TimvdLippe commented:
@bicknellr I am not sure if I follow your previous comment RE the issue in #141. Could you provide a test case example that shows the issue?

TimvdLippe commented:
Oh I think I understand now. Taking a stab at it.

TimvdLippe committed:
Fix incorrect upgrading of siblings when setting outerHTML

TimvdLippe commented:
All right, problem solved :tada:

TimvdLippe committed:
Add missing assert for both siblings

TimvdLippe committed:
Make null-comparison explicit

This faster for V8 to compare

TimvdLippe committed:
Revert "Make null-comparison explicit"

This reverts commit b5a11656064af02a5c43480aa5e4d02d037acc2f.

bicknellr commented on src/Patch/Element.js:
```diff
@@ -116,6 +128,53 @@ export default function(internals) {
+          patchFunction.call(internals, sibling);
+          sibling = sibling.nextSibling;
+        }
+      }
```
I think this needs another case at the beginning that checks if `parentNode === null`: in that case, there's no need to call `patchFunction` with `parentNode` (which would otherwise happen because `previousSibling === null && nextSibling === null`).

TimvdLippe committed:
Handle more corner cases

- Throw error when parentNode does not exist
- Fix namespace creation element to be compatible with SVGElement

TimvdLippe committed:
Skip SVGElement test on Safari 10

TimvdLippe committed:
Fix skip logic for Edge 14

TimvdLippe commented:
@bicknellr I added the SVG element tests. PTAL

dfreedm commented:
Remaking PR in the monorepo!
